sklearn_numba_dpex/common/random.py

```diff
@@ -0,0 +1,307 @@
+# Reference resource about PRNG: https://prng.di.unimi.it/
+
+# NB2: original numba.cuda.random code also includes functions for generating normally
+# distributed floats but we don't include it here as long as it's not needed.
+
+zero_idx = int64(0)
+one_idx = int64(1)
+
+
+def get_random_raw(states):
+    result = dpt.empty(sh=(1,), dtype=np.uint64, device=states.device)
+    make_random_raw_kernel()(states, result)
+    return result
+
+
+@lru_cache
+def make_random_raw_kernel():
+    @dpex.kernel
+    def _get_random_raw_kernel(states, result):
+        result[zero_idx] = _xoroshiro128pp_next(states, zero_idx)
+
+    return _get_random_raw_kernel[1, 1]
+
+
+@lru_cache
+def make_rand_uniform_kernel_func(dtype):
+    if not hasattr(dtype, "name"):
+        raise ValueError(
+            "dtype is expected to have an attribute 'name', like np.dtype or numba "
+            "types."
+        )
+
+    if dtype.name == "float64":
+        convert_rshift = uint32(11)
+        convert_const = uint64(1) << uint32(53)
+        convert_const_one = float64(1)
+
+        @dpex.func
+        def uint64_to_unit_float(x):
+            """Convert uint64 to float64 value in the range [0.0, 1.0)"""
+            return (x >> convert_rshift) * (convert_const_one / convert_const)
+
+    elif dtype.name == "float32":
+        convert_rshift = uint32(40)
+        convert_const = float32(uint32(1) << uint32(24))
+        convert_const_one = float32(1)
+
+        @dpex.func
+        def uint64_to_unit_float(x):
+            """Convert uint64 to float32 value in the range [0.0, 1.0)
+            NB: this is different than original numba.cuda.random code. Instead of
+            generating a float64 random number before casting it to float32, a float32
+            number is generated from uint64 without intermediate float64. This enables
+            compatibility with devices that do not support float64 numbers.
+            However is seems to be exactly equivalent e.g it passes the float precision
+            test in sklearn.
+            """
+            return float32(x >> convert_rshift) * (convert_const_one / convert_const)
+
+    else:
+        raise ValueError(
+            "Expected dtype.name in {float32, float64} but got "
+            f"dtype.name == {dtype.name}"
+        )
+
+    @dpex.func
+    def xoroshiro128pp_uniform(states, index):
+        return uint64_to_unit_float(_xoroshiro128pp_next(states, index))
+
+    return xoroshiro128pp_uniform
+
+
+def create_xoroshiro128pp_states(n_states, seed=None, subsequence_start=0, device=None):
+    """Returns a new device array initialized for n random number generators.
+
+    This initializes the RNG states so that each state in the array corresponds
+    subsequences in the separated by 2**64 steps from each other in the main
+    sequence.  Therefore, as long no thread requests more than 2**64 random numbers,
+    all of the RNG states produced by this function are guaranteed to be independent.
```

Comment:
```suggestion
    This initializes the RNG states so that states in the array corresponds
    to subsequences the separated by 2**64 steps from each other in the main
    sequence.
    Therefore, as long as no thread requests more than 2**64 random numbers,
    all the RNG states produced by this function are guaranteed to be independent.
```